File: services/tests/servicestests/src/com/android/server/integrity/utils/TestUtils.java
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.server.integrity.utils;

public class TestUtils {

    public static String getBits(long component, int numOfBits) {
        return String.format("%" + numOfBits + "s", Long.toBinaryString(component))
                .replace(' ', '0');
    }

    public static String getValueBits(String value) {
        StringBuilder stringBuilder = new StringBuilder();
        for (byte valueByte : value.getBytes()) {
            stringBuilder.append(getBits(valueByte, /* numOfBits= */ 8));
        }
        return stringBuilder.toString();
    }

    public static byte[] getBytes(String bits) {
        int bitStringSize = bits.length();
        int numOfBytes = bitStringSize / 8;
        if (bitStringSize % 8 != 0) {
            numOfBytes++;
        }
        byte[] bytes = new byte[numOfBytes];
        for (int i = 0; i < bits.length(); i++) {
            if (bits.charAt(i) == '1') {
                bytes[i / 8] |= 1 << (7 - (i % 8));
            }
        }
        return bytes;
    }
}

File: services/tests/displayservicetests/src/com/android/server/display/brightness/TestUtils.kt
/*
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.server.display.brightness

import com.android.server.display.HysteresisLevels
import com.android.server.display.config.SensorData

@JvmOverloads
fun createLightSensorControllerConfig(
    initialSensorRate: Int = 1,
    normalSensorRate: Int = 2,
    resetAmbientLuxAfterWarmUpConfig: Boolean = true,
    ambientLightHorizonShort: Int = 1,
    ambientLightHorizonLong: Int = 10_000,
    lightSensorWarmUpTimeConfig: Int = 0,
    weightingIntercept: Int = 10_000,
    ambientBrightnessThresholds: HysteresisLevels = createHysteresisLevels(),
    ambientBrightnessThresholdsIdle: HysteresisLevels = createHysteresisLevels(),
    brighteningLightDebounceConfig: Long = 100_000,
    darkeningLightDebounceConfig: Long = 100_000,
    brighteningLightDebounceConfigIdle: Long = 100_000,
    darkeningLightDebounceConfigIdle: Long = 100_000,
    ambientLightSensor: SensorData = SensorData()
) = LightSensorController.LightSensorControllerConfig(
    initialSensorRate,
    normalSensorRate,
    resetAmbientLuxAfterWarmUpConfig,
    ambientLightHorizonShort,
    ambientLightHorizonLong,
    lightSensorWarmUpTimeConfig,
    weightingIntercept,
    ambientBrightnessThresholds,
    ambientBrightnessThresholdsIdle,
    brighteningLightDebounceConfig,
    darkeningLightDebounceConfig,
    brighteningLightDebounceConfigIdle,
    darkeningLightDebounceConfigIdle,
    ambientLightSensor
)

fun createHysteresisLevels(
    brighteningThresholdsPercentages: FloatArray = floatArrayOf(),
    darkeningThresholdsPercentages: FloatArray = floatArrayOf(),
    brighteningThresholdLevels: FloatArray = floatArrayOf(),
    darkeningThresholdLevels: FloatArray = floatArrayOf(),
    minDarkeningThreshold: Float = 0f,
    minBrighteningThreshold: Float = 0f,
    potentialOldBrightnessRange: Boolean = false
) = HysteresisLevels(
    brighteningThresholdsPercentages,
    darkeningThresholdsPercentages,
    brighteningThresholdLevels,
    darkeningThresholdLevels,
    minDarkeningThreshold,
    minBrighteningThreshold,
    potentialOldBrightnessRange
)